analyze-hotspots/src/main/java/com/esri/arcgismaps/sample/analyzehotspots/components/MapViewModel.kt

```diff
@@ -0,0 +1,171 @@
+
+        val queryString = StringBuilder("(\"DATE\" > date '")
+            .append(fromDate)
+            .append(" 00:00:00' AND \"DATE\" < date '")
+            .append(toDate)
+            .append(" 00:00:00')")
+
+        //geoprocessingParameters.inputs["Query"] = geoprocessingString
+        geoprocessingParameters.inputs["Query"] = GeoprocessingString(queryString.toString())
+
+        // create and start geoprocessing job
+        geoprocessingJob = geoprocessingTask.createJob(geoprocessingParameters)
+
+        runGeoprocessingJob()
+    }
+
+    /**
+     * Starts the [geoprocessingJob], shows the progress dialog and
+     * displays the result hotspot map image layer to the MapView
+     */
+    private suspend fun runGeoprocessingJob() {
+        // display the progress dialog
+        showJobProgressDialog.value = true
+        // start the job
+        geoprocessingJob?.start()
+        // collect the job progress
+        sampleCoroutineScope.launch {
+            geoprocessingJob?.progress?.collect { progress ->
+                // updates the job progress dialog
+                geoprocessingJobProgress.value = progress
+            }
+        }
+        // get the result of the job on completion
+        geoprocessingJob?.result()?.onSuccess {
+            // dismiss the progress dialog
+            showJobProgressDialog.value = false
+            // get the job's result
+            val geoprocessingResult = geoprocessingJob?.result()?.getOrElse {
+                showErrorDialog(it.message.toString(), it.cause.toString())
+            } as GeoprocessingResult
+            // resulted hotspot map image layer
+            val hotspotMapImageLayer = geoprocessingResult.mapImageLayer?.apply {
+                opacity = 0.5f
+            } ?: return showErrorDialog("Result map image layer is null")
+
+            // add new layer to map
+            mapViewState.value.arcGISMap.operationalLayers.add(hotspotMapImageLayer)
+        }?.onFailure { throwable ->
+            showErrorDialog(throwable.message.toString(), throwable.cause.toString())
+            showJobProgressDialog.value = false
+        }
+    }
+
+    fun cancelGeoprocessingJob(coroutineScope: CoroutineScope) {
+        coroutineScope.launch {
+            geoprocessingJob?.cancel()
+        }
+    }
+
+    /**
+     * Convert epoch time in [millis] to String date format
+     */
+    fun convertMillisToString(millis: Long): String {
+        val instant = Instant.ofEpochMilli(millis)
+        val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
+        val date = LocalDateTime.ofInstant(instant, ZoneId.systemDefault())
+        return date.format(formatter)
+    }
+
+    // error dialog status
+    val errorDialogStatus = mutableStateOf(false)
+    var errorTitle = ""
+    var errorDescription = ""
+
+    /**
+     * Displays an error dialog with [title] and optional [description]
+     */
+    fun showErrorDialog(title: String, description: String = "") {
```

Comment:
It looks like the `showErrorDialog` fun will be used by other samples too. Should we consider moving it to common samples-lib?

Reply:
Good point, added a `MessageDialogViewModel` to the `MessageDialog` component in the samples-lib